README.md
# OpenFHE integer examples

Sample programs for Encrypted Integer Processing

# Encrypted substring search

Based on the [Rabin-Karp algorithm](https://en.wikipedia.org/wiki/Rabin–Karp_algorithm)

Source code adapted from [C++ source code published here](https://www.sanfoundry.com/cpp-program-implement-rabin-karp-method-for-string-matching/)

> src/strsearch_enc_1.cpp encrypted string search no SIMD batching

> src/strsearch_enc_2.cpp encrypted string search with SIMD batching

# Building

Building the system
===================

Build instructions for Ubuntu
---------

Please note that we have not tried installing this on windows or
macOS. If anyone does try this, please update this file with
instructions.  It's recommended to use at least Ubuntu 18.04, and gnu g++ 7 or greater.


1. Install pre-requisites (if not already installed):
`g++`, `cmake`, `make`, and `autoconf`. Sample commands using `apt-get` are listed below. It is possible that these are already installed on your system.


```bash
sudo apt-get install build-essential #this already includes g++
sudo apt-get install autoconf
sudo apt-get install make
sudo apt-get install cmake
```

> Note that `sudo apt-get install g++-<version>` can be used to
install a specific version of the compiler. You can use `g++
--version` to check the version of `g++` that is the current system
default.

2. Install OpenFHE on your system. This code was tested with pre-release 1.10.3.


- instructions can be found on the [official installation documentation](https://openfhe-development.readthedocs.io/en/latest/sphinx_rsts/intro/installation/installation.html)

Run `make install` at the end to install the system to the default
location (you can change this location, but then you will have to
change the Makefile in this repo to reflect the new location).

3. Clone this repo onto your system.

4. Create the bin directory

```
mkdir build
```

5. Build the system using cmake

```
cd build
cmake ..
make
```

All the examples will be in the `build/bin` directory. All input files, and resulting assembler outputs will be in various subdirectories under `examples`.

Running Simple Examples
=======================

From the root directory, run the two string search examples with 

> `bin/strsearch_enc_1`

> `bin/strsearch_enc_2`




Release_Notes.md
09/30/2020: OpenFHE encrypted substring search v1.0 is released

* Initial prototype release 
* Tested with OpenFHE v10.3

Comments/additions are welcome by the OpenFHE team


src/debug.h
// @file debug.h This file contains macros and associated helper functions for
// quick cerr oriented debugging that can be quickly enabled and disabled. It
// also contains functions for timing code.
// @author TPOC: contact@palisade-crypto.org
//
// @copyright Copyright (c) 2019, New Jersey Institute of Technology (NJIT)
// All rights reserved.
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
// 1. Redistributions of source code must retain the above copyright notice,
// this list of conditions and the following disclaimer.
// 2. Redistributions in binary form must reproduce the above copyright notice,
// this list of conditions and the following disclaimer in the documentation
// and/or other materials provided with the distribution. THIS SOFTWARE IS
// PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
// IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
// EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
// INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
// ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

#ifndef __dbg_h__
#define __dbg_h__

// include <iostream>
// include <cstdlib.h>

/* defining NDEBUG in the compile line turns everything off.
   unless PROFILE is defined in the file before all includes,'
   in which case TIC/TOC will still work and PROFILELOG() can be
   used for logging results to std::cout, and DEBUG() will remain
   silent. dbg_flag does not get used by PROFILELOG()
 */

#include <time.h>
#include <chrono>  // for timing
#include <utility>

#include "utils/inttypes.h"

#if !defined(NDEBUG)

// note that for the following dbg_flag needs to be defined in some scope using
// DEBUG_FLAG
#define DEBUG_FLAG(x) bool dbg_flag = x;

// debugging macro prints value of x on cerr
#define DEBUG(x)                   \
  do {                             \
    if (dbg_flag) {                \
      std::cerr << x << std::endl; \
    }                              \
  } while (0)

// debugging macro prints typography of x and value of x on cerr
#define DEBUGEXP(x)                             \
  do {                                          \
    if (dbg_flag) {                             \
      std::cerr << #x << ":" << x << std::endl; \
    }                                           \
  } while (0)

// debugging macro prints value of x and location in codex on cerr
#define DEBUGWHERE(x)                                               \
  do {                                                              \
    if (dbg_flag) {                                                 \
      std::cerr << #x << ":" << x << " at " << __FILE__ << " line " \
                << __LINE__ << std::endl;                           \
    }                                                               \
  } while (0)

#if defined(PROFILE)  // Profiler works

#define PROFILELOG(x)              \
  do {                             \
    if (true) {                    \
      std::cout << x << std::endl; \
    }                              \
  } while (0)

// debugging macro prints typography of x and value of x on cerr
#define PROFILELOGEXP(x)                        \
  do {                                          \
    if (true) {                                 \
      std::cout << #x << ":" << x << std::endl; \
    }                                           \
  } while (0)

// debugging macro prints value of x and location in codex on cerr
#define PROFILELOGWHERE(x)                                          \
  do {                                                              \
    if (true) {                                                     \
      std::cout << #x << ":" << x << " at " << __FILE__ << " line " \
                << __LINE__ << std::endl;                           \
    }                                                               \
  } while (0)

#else  // #if!defined(PROFILE) // profiling a noop
#define PROFILELOG(x)
#define PROFILELOGEXP(x)
#define PROFILELOGWHERE(x)

#endif  // PROFILE

#define TIC(t) t = timeNow()
#define TOC(t) duration(timeNow() - t)
#define TOC_NS(t) duration_ns(timeNow() - t)
#define TOC_US(t) duration_us(timeNow() - t)
#define TOC_MS(t) duration_ms(timeNow() - t)
#define TOC_S(t) duration_s(timeNow() - t)

#else  // NDEBUG

// #define debug(M, ...) fprintf(stderr, "DEBUG %s:%d: " M "\n", __FILE__,
// __LINE__, ##__VA_ARGS__)

#if !defined(PROFILE)

// these are turned off functions

#define DEBUG_FLAG(x)
#define DEBUG(x)
#define DEBUGEXP(x)
#define DEBUGWHERE(x)

#define PROFILELOG(x)
#define PROFILELOGEXP(x)
#define PROFILELOGWHERE(x)

#define TIC(t) t = timeNow()
#define TOC(t) std::chrono::steady_clock::duration::zero().count()
#define TOC_NS(t) std::chrono::steady_clock::duration::zero().count()
#define TOC_US(t) std::chrono::steady_clock::duration::zero().count()
#define TOC_MS(t) std::chrono::steady_clock::duration::zero().count()
#define TOC_S(t) std::chrono::steady_clock::duration::zero().count()

#else  // PROFILE
// if PROFILE is turned on, then TIC TOC still work and

#define DEBUG_FLAG(x)
#define DEBUG(x)
#define DEBUGEXP(x)
#define DEBUGWHERE(x)

#define PROFILELOG(x)              \
  do {                             \
    if (true) {                    \
      std::cerr << x << std::endl; \
    }                              \
  } while (0)

// debugging macro prints typography of x and value of x on cerr
#define PROFILELOGEXP(x)                        \
  do {                                          \
    if (true) {                                 \
      std::cout << #x << ":" << x << std::endl; \
    }                                           \
  } while (0)

// debugging macro prints value of x and location in codex on cerr
#define PROFILELOGWHERE(x)                                          \
  do {                                                              \
    if (true) {                                                     \
      std::cout << #x << ":" << x << " at " << __FILE__ << " line " \
                << __LINE__ LL std::endl;                           \
    }                                                               \
  } while (0)

#define TIC(t) t = timeNow()
#define TOC(t) duration(timeNow() - t)
#define TOC_NS(t) duration_ns(timeNow() - t)
#define TOC_US(t) duration_us(timeNow() - t)
#define TOC_MS(t) duration_ms(timeNow() - t)
#define TOC_S(t) duration_s(timeNow() - t)

#endif  // PROFILE

#endif  // NDEBUG

typedef std::chrono::high_resolution_clock::time_point TimeVar;

#define duration(a) \
  std::chrono::duration_cast<std::chrono::milliseconds>(a).count()
#define duration_ns(a) \
  std::chrono::duration_cast<std::chrono::nanoseconds>(a).count()
#define duration_us(a) \
  std::chrono::duration_cast<std::chrono::microseconds>(a).count()
#define duration_ms(a) \
  std::chrono::duration_cast<std::chrono::milliseconds>(a).count()
#define duration_s(a) \
  std::chrono::duration_cast<std::chrono::seconds>(a).count()
#define timeNow() std::chrono::high_resolution_clock::now()

double currentDateTime();

template <typename F, typename... Args>
double funcTime(F func, Args&&... args) {
  TimeVar t1 = timeNow();
  func(std::forward<Args>(args)...);
  return duration(timeNow() - t1);
}

#endif  // #__dbg_h__


src/ss_enc_1.cpp
/* PALISADE C++ program implements the Rabin-Karp method for string matching using
 * encrypted computation
 * plaintext version of this code comes from 
 * https://www.sanfoundry.com/cpp-program-implement-rabin-karp-method-for-string-matching
 * author David Bruce Cousins@dualitytech.com

 */


#include <stdio.h>
#include <string.h>
#include <iostream>
#include <fstream>
#include <vector>
#include "debug.h"
#include "palisade.h"
using namespace std;
using namespace lbcrypto;

//data types we will need
using CT = Ciphertext<DCRTPoly> ; //ciphertext
using PT = Plaintext ; //plaintext
using vecCT = vector<CT>; //vector of ciphertexts
using vecPT = vector<PT>; //vector of plaintexts
using vecInt = vector<int64_t>; // vector of ints
using vecChar = vector<char>; // vector of characters

// d is the number of characters in input alphabet
const int d = 256;
     
/*  pat  -> pattern
	txt  -> text
	p    -> A prime number
*/
     
void get_input_from_term(vecChar& a) {
  string cstr;
  cin.ignore(numeric_limits<streamsize>::max(),'\n'); //flushes buffer
  std::getline(std::cin, cstr);
  cout <<"Pattern is `"<<cstr<<"'"<<endl;    
  for(auto c: cstr) {
	a.push_back(c);
  }
  cout <<"Pattern is "<<a.size()<<" characters"<<endl;  
  return;
}
     
void get_input_from_file(vecChar& a, string fname) {
  char c;

  ifstream in_file;
  in_file.open(fname);
  if (!in_file) {
	cerr << "Can't open file for input: "<<fname;
	exit(-1); //error exit
  }
  
  while (in_file >> c) {
	a.push_back(c);
  }
  cout <<"Read "<<a.size()<<" characters"<<endl;
  in_file.close();
  return;
}
     
     
void search(vecChar &pat, vecChar &txt, int ps) {
  long p(ps);
  DEBUG_FLAG(true);
  int M = pat.size();
  DEBUGEXP(M);
  int N = txt.size();
  DEBUGEXP(N);
  int i, j;
  long ph = 0;  // hash value for pattern
  long th = 0; // hash value for txt
  long h = 1;

  int nfound = 0;
     
  // The value of h would be "pow(d, M-1)%p"
  for (i = 0; i < M-1; i++) {
	h = (h*d)%p;
  }
  // Calculate the hash value of pattern and first window of text
  for (i = 0; i < M; i++) {
	ph = (d * ph + pat[i]) % p;
	th = (d * th + txt[i]) % p;
  }
     
  // Slide the pattern over text one by one
  for (i = 0; i <= N - M; i++) {
	
	// Check the hash values of current window of text and pattern
	// If the hash values match then only check for characters on by one
	if ( ph == th )	{
	  /* Check for characters one by one */
	  for (j = 0; j < M; j++) {
		if (txt[i + j] != pat[j])
		  break;
	  }
	  if (j == M) { // if ph == t and pat[0...M-1] = txt[i, i+1, ...i+M-1]

		cout<<"Pattern found at index "<< i << endl;
		nfound++;
	  }
	}
     
	// Calculate hash value for next window of text: Remove leading digit,
	// add trailing digit
	if ( i < N - M ) {
	  th = (d * (th - txt[i] * h) + txt[i + M]) % p;
     
	  // We might get negative value of t, converting it to positive
	  if (th < 0) {
		th = (th + p);
	  }
	}
  } //end for
  cout<<"total occurances " <<nfound<<endl;
}
     

CT encrypt_repeated_integer(CryptoContext<DCRTPoly> &cc, LPPublicKey<DCRTPoly> &pk, int64_t in, size_t n){
  
  vecInt v_in(0);
  for (auto i = 0; i < n; i++){
	v_in.push_back(0);
  }
  PT pt= cc->MakePackedPlaintext(v_in);
  CT ct = cc->Encrypt(pk, pt);
  return ct;
}	
     
vecCT encrypted_search(CryptoContext<DCRTPoly> &cc,  LPPublicKey<DCRTPoly> &pk, vecCT &epat, vecCT &etxt) {
  DEBUG_FLAG(true);
  int M = epat.size();
  DEBUGEXP(M);
  int N = etxt.size();
  DEBUGEXP(N);
  int i, j;

  size_t nrep(1);

  CT phct = encrypt_repeated_integer(cc, pk, 0, nrep);  // hash value for pattern
  CT thct = encrypt_repeated_integer(cc, pk, 0, nrep);  // hash value for txt
  CT hct = encrypt_repeated_integer(cc, pk, 1, nrep);  // encrypted h
  CT dct = encrypt_repeated_integer(cc, pk, d, nrep);  // encrypted d

  int nfound = 0;
     
  // The value of h would be "pow(d, M-1)%p"
  for (i = 0; i < M-1; i++) {
	hct = cc->ComposedEvalMult(hct, dct);
  }

  // Calculate the hash value of pattern and first window of text
  for (i = 0; i < M; i++) {
	auto tmp = cc->EvalAdd(phct,epat[i]);
	phct = cc->ComposedEvalMult(tmp, dct);
	tmp = cc->EvalAdd(thct,etxt[i]);
	thct = cc->ComposedEvalMult(tmp, dct);
  }

  vecCT eresult(0);
  // Slide the pattern over text one by one
  for (i = 0; i <= N - M; i++) {
	cout<<i<< '\r'<<flush;
	
	// Check the hash values of current window of text and pattern
	// If the hash values match then only check for characters on by one
	// subtract the two hashes, zero is equality

	eresult.push_back(cc->EvalSub(phct, thct));
     
	// Calculate hash value for next window of text: Remove leading digit,
	// add trailing digit
	if ( i < N - M ) {
	  
	  //th = (d * (th - txt[i] * h) + txt[i + M]) % p;
	  thct = cc->EvalAdd(cc->ComposedEvalMult(dct,
											  cc->EvalSub(thct,
														  cc->ComposedEvalMult(etxt[i], hct)
														  )
											  ), etxt[i+M]
						 );
	}
  } //end for
  return eresult;
}
     
int main()
{
  vecChar txt;
  vecChar pat;
  string infilename;

  cout<<sizeof(int)<<endl;
  cout<<sizeof(long)<<endl;
  
  cout<<"Enter file for Text:";
  //cin >> infilename;
  //infilename = "data/alice.txt";
  infilename = "data/warandpeace.txt";
  get_input_from_file(txt, infilename);

  cout<<"Enter Pattern to Search:";
  get_input_from_term(pat);

  int p = 786433; //plaintext prime modulus

  cout<<"p "<<p<<endl;
  TIC(auto t1);
  search(pat, txt, p);
  auto plain_time_ms = TOC_MS(t1);
  cout<< "Plaintext execution time "<<plain_time_ms<<" mSec."<<endl;

  cout <<"setting up BGV RNS crypto system"<<endl;
  

  uint32_t plaintextModulus = p;
  uint32_t multDepth = 50;
  double sigma = 3.2;
  SecurityLevel securityLevel = HEStd_128_classic;
  
  CryptoContext<DCRTPoly> cc =
      CryptoContextFactory<DCRTPoly>::genCryptoContextBGVrns(
          multDepth, plaintextModulus, securityLevel, sigma, 2, OPTIMIZED, HYBRID, 0, 0, 0, 0, 0, 0, MANUAL);
	
  // Enable features that you wish to use
  cc->Enable(ENCRYPTION);
  cc->Enable(SHE);
  cc->Enable(LEVELEDSHE);

  cout<<"Step 2 - Key Generation"<<endl;
  
  // Initialize Public Key Containers
  // Generate a public/private key pair
  auto keyPair = cc->KeyGen();
  
  // Generate the relinearization key
  cc->EvalMultKeyGen(keyPair.secretKey);
  
  cout<<"Step 3 - Encryption"<<endl;  

  auto ringsize = cc->GetRingDimension();
  cout << "ringsize = "<<ringsize << endl;
									 

  //encrypt the pattern
  vecInt vin(0);
  vecCT epat(0);
  unsigned int j(0);
  for (auto ch: pat) {
	cout<<j<< '\r'<<flush;
	j++;
	vin.push_back(ch);
	PT pt= cc->MakePackedPlaintext(vin);
	vin.clear();
	CT ct = cc->Encrypt(keyPair.publicKey, pt);
	epat.push_back(ct);
  }	

  //encrypt the text
  unsigned int nbatch = ceil(txt.size()/ringsize);
  cout << "can store "<<nbatch <<" batches in the ct"<<endl;
	
  //for now lets only do the first 64 characters

  vecCT etxt(0);
  auto pt_len(0);
  for (auto i = 0; i < 64; i++) {
	cout<<i<< '\r'<<flush;

	vin.push_back(txt[i]);
	Plaintext pt= cc->MakePackedPlaintext(vin);
	pt_len = pt->GetLength();
	vin.clear();
	CT ct = cc->Encrypt(keyPair.publicKey, pt);
	etxt.push_back(ct);
  }	
  cout<<"Step 4 - Encrypted string search"<<endl;  

  TIC(auto t2);
  vecCT eresult = encrypted_search(cc, keyPair.publicKey, epat, etxt);
  auto encrypted_time_ms = TOC_MS(t2);
		  cout<< "Encrypted execution time "<<encrypted_time_ms<<" mSec."<<endl;  

  


  vecPT vecResult(0);
  for (auto e_itr:eresult){
	PT ptresult;  
	cc->Decrypt(keyPair.secretKey, e_itr, &ptresult);
	ptresult->SetLength(pt_len);
	vecResult.push_back(ptresult);
  }

  cout<<"Result "<<vecResult << endl;
  return 0;
}


src/strsearch_enc_2.cpp
/* OpenFHE C++ program implements the Rabin-Karp method for string
 * matching using encrypted computation with SIMD batching
 * plaintext version of this code comes from 
 * https://www.sanfoundry.com/cpp-program-implement-rabin-karp-method-for-string-matching
 * author David Bruce Cousins@dualitytech.com

 */


#include <cstdio>
#include <cstring>
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

#include "openfhe.h"
#include "utils/debug.h"
#include "scheme/bfvrns/cryptocontext-bfvrns.h"
#include "gen-cryptocontext.h"
using namespace std;
using namespace lbcrypto;

//data types we will need
using CT = Ciphertext<DCRTPoly> ; //ciphertext
using PT = Plaintext ; //plaintext
using vecCT = vector<CT>; //vector of ciphertexts
using vecPT = vector<PT>; //vector of plaintexts
using vecInt = vector<int64_t>; // vector of ints
using vecChar = vector<char>; // vector of characters

// d is the number of characters in input alphabet
const int32_t d = 256; 
     
/*  pat  -> pattern
	txt  -> text
	p    -> A prime number
*/


// function to get string input from terminal and return as vector of char
void get_input_from_term(vecChar& a) {
  string cstr;
  cin.ignore(numeric_limits<streamsize>::max(),'\n'); //flushes buffer
  cin >> ws; //discards white space
  std::getline(std::cin, cstr);
  cout <<"Pattern is `"<<cstr<<"'"<<endl;    
  for(auto c: cstr) {
	  a.push_back(c);
  }
  cout <<"Pattern is "<<a.size()<<" characters"<<endl;  
  return;
}

// function to read text from a file and return as vector of char
void get_input_from_file(vecChar& a, string fname) {
  char c;

  ifstream in_file;
  in_file.open(fname);
  if (!in_file) {
    cerr << "Can't open file for input: "<<fname<<endl;
    exit(-1); //error exit
  }
  
  while (in_file >> c) {
	a.push_back(c);
  }
  cout <<"Read "<<a.size()<<" characters"<<endl;
  in_file.close();
  return;
}
     
// plaintext string search of pat within txt, with modulus of ps     
vecInt search(vecChar &pat, vecChar &txt, int ps) {
  int64_t p(ps);
  DEBUG_FLAG(false);
  size_t M = pat.size();
  DEBUGEXP(M);
  size_t N = txt.size();
  DEBUGEXP(N);
  size_t i, j;
  int64_t ph = 0;  // hash value for pattern
  int64_t th = 0; // hash value for txt
  int64_t h = 1;

  int nfound = 0;
     
  // The value of h would be "pow(d, M-1)%p"
  for (i = 0; i < M-1; i++) {
    h = (h*d)%p;
    DEBUGEXP(h);
  }
  DEBUG(" hfinal: "<<h);

  // Calculate the hash value of pattern and first window of text
  for (i = 0; i < M; i++) {
    ph = (d * ph + pat[i]) % p;
    th = (d * th + txt[i]) % p;
  }
  DEBUG(" initial ph: "<<ph);
  DEBUG(" initial th: "<<th);
  vecInt pres(0);
  // Slide the pattern over text one by one
  for (i = 0; i <= N - M; i++) {
	
	// Check the hash values of current window of text and pattern
	// If the hash values match then only check for characters on by one
	if ( ph == th )	{
	  /* Check for characters one by one */
	  for (j = 0; j < M; j++) {
		if (txt[i + j] != pat[j])
		  break;
	  }
	  if (j == M) { // if ph == t and pat[0...M-1] = txt[i, i+1, ...i+M-1]

		//cout<<"Pattern found at index "<< i << endl;
		pres.push_back(i);
		nfound++;
	  }
	}
     
	// Calculate hash value for next window of text: Remove leading digit,
	// add trailing digit
	if ( i < N - M ) {
	  th = (d * (th - txt[i] * h) + txt[i + M]) % p;

	  // We might get negative value of t, converting it to positive
	  if (th < 0) {
		th = (th + p);
	  }

	  //cout<<" "<<th;
	}
  } //end for
  //cout<<endl;
  cout<<"total occurences " <<nfound<<endl;
  return pres;
}
     
// helper function to encrypt an integer repeatedly into a packed plaintext
// and encrypt it
CT encrypt_repeated_integer(CryptoContext<DCRTPoly> &cc, PublicKey<DCRTPoly> &pk,  int64_t in, size_t n){
  
  vecInt v_in(n, in);
  PT pt= cc->MakePackedPlaintext(v_in);
  CT ct = cc->Encrypt(pk, pt);

  return ct;
}

// helper function to multiply by constant 256 using binary tree addition
CT encMultD(CryptoContext<DCRTPoly> &cc, CT in){
  if (d !=256){
	cout <<"error d not 256"<<endl;
	exit(-1);
  }
  auto tmp(in);
  for (size_t i = 0; i< 8; i++ ){
	  tmp = cc->EvalAdd(tmp, tmp);
  }
  
  return(tmp);
}

//SIMD encrypted search
vecCT encrypted_search(CryptoContext<DCRTPoly> &cc,  PublicKey<DCRTPoly> &pk,  vecCT &epat, vecCT &etxt, int ps) {

  int64_t p(ps);
  DEBUG_FLAG(false);
  size_t M = epat.size();
  DEBUGEXP(M);
  size_t N = etxt.size();
  DEBUGEXP(N);
  size_t i;

  PT dummy;
  
  size_t nrep(cc->GetRingDimension());
  DEBUG("encrypting small ct");
  CT phct = encrypt_repeated_integer(cc, pk, 0, nrep);  // hash value for pattern
  CT thct = encrypt_repeated_integer(cc, pk, 0, nrep);  // hash value for txt

  CT dhct = encrypt_repeated_integer(cc, pk, d, nrep);  // d

  DEBUG("encrypting hct");     
  // The value of h would be "pow(d, M-1)%p"
  long h = 1;
  for (i = 0; i < M-1; i++) {
	  h = (h*d)%p;
  }
  CT hct = encrypt_repeated_integer(cc, pk, h, nrep);  // encrypted h

  DEBUG("encrypting first hashes" );     
  // Calculate the hash value of pattern and first window of text
  for (i = 0; i < M; i++) {
    auto tmp = encMultD(cc, phct);	
    phct = cc->EvalAdd(tmp, epat[i]);

    tmp = encMultD(cc, thct);
    thct = cc->EvalAdd(tmp, etxt[i]);
  }

  vecCT eres(0);
  // Slide the pattern over text one by one
  DEBUG("sliding" );     
  for (i = 0; i <= N - M; i++) {
    cout<<i<< '\r'<<flush;
    
    // Check the hash values of current window of text and pattern
    // If the hash values match then only check for characters on by one
    // subtract the two hashes, zero is equality
    DEBUG("sub" );     
    eres.push_back(cc->EvalSub(phct, thct));
      
    // Calculate hash value for next window of text: Remove leading digit,
    // add trailing digit
    if ( i < N - M ) {
      DEBUG("rehash" );     
      //th = (d * (th - txt[i] * h) + txt[i + M]) % p;
      auto tmp = encMultD(cc,
                cc->EvalSub(thct,
                      cc->EvalMult(etxt[i], hct)
                      )
                );

      thct = cc->EvalAdd(tmp, etxt[i+M] );
    }

  } //end for
  return eres;
}
     
int main()
{
  vecChar txt;
  vecChar pat;
  string infilename;

  //cout<<"Enter file for Text:";
  //cin >> infilename;
  infilename = "data/annakarenina.txt";
  get_input_from_file(txt, infilename);

  //cout<<"Enter buffer size:";
  uint32_t maxNBatches(0);
  uint32_t minNBatches(0);
  //cin>> maxNBatches;
  maxNBatches = 32;
  minNBatches = 10;
  cout << "batching to "<<maxNBatches<< " characters max"<<endl;

  //cout<<"Enter Pattern to Search:";
  
  //get_input_from_term(pat);
  pat = {'A', 'n', 'n', 'a'};

  int32_t p = 786433; //plaintext prime modulus
  //int p = 65537; // use this to show core dump

  cout<<"p "<<p<<endl;
  TIC(auto t1);

  auto presult = search(pat, txt, p);
  auto plain_time_ms = TOC_MS(t1);
  cout<< "Plaintext execution time "<<plain_time_ms<<" mSec."<<endl;

  cout <<"setting up BFV RNS crypto system"<<endl;

  uint32_t plaintextModulus = p;
  uint32_t multDepth = 32; //works for 32 batches

  double sigma = 3.2;
  SecurityLevel securityLevel = HEStd_128_classic;

  lbcrypto::CCParams<lbcrypto::CryptoContextBFVRNS> parameters;
  parameters.SetPlaintextModulus(plaintextModulus);
  parameters.SetSecurityLevel(securityLevel);
  parameters.SetStandardDeviation(sigma);
  parameters.SetMultiplicativeDepth(multDepth);

  lbcrypto::CryptoContext<lbcrypto::DCRTPoly > cc = lbcrypto::GenCryptoContext(parameters);

    cc->Enable(PKE);
    cc->Enable(KEYSWITCH);
    cc->Enable(LEVELEDSHE);


  cout<<"Step 2 - Key Generation"<<endl;
  
  // Initialize Public Key Containers
  // Generate a public/private key pair
  auto keyPair = cc->KeyGen();
  
  // Generate the relinearization key
  cc->EvalMultKeyGen(keyPair.secretKey);

  // note we do not use rotation in this example so we don't need rotation keys
  // but make them anyway 
  // Generate the rotation evaluation keys
  //cc->EvalAtIndexKeyGen(keyPair.secretKey, {1, 2, -1, -2}); 

  cout<<"Step 3 - Encryption"<<endl;  

  cout<<"Step 3.1 - Encrypt text"<<endl;  

  //encrypt the text
  auto ringsize = cc->GetRingDimension();
  cout << "Given ringsize = "<<ringsize << endl;
  cout << "and text size = "<<txt.size() << endl;
  uint32_t nbatchEst = int(ceil(float(txt.size())/float(ringsize)));
  cout << "We can store approximately "<<nbatchEst <<" batches in the CT"<<endl;

  //need to encrypt the text in batches, each one consists of
  //nbatch points. the next batch starts at nbatch-M 
  //so each CT in the vecCT has ringsize entries.
  // vecCT[0] has characters {0, nbatch-M, 2(nbatch-M) etc..}
  // vecCT[1] has characters {1, 1+(nbatch-M), 1+2(nbatch-M) etc..}
  cout << "Adjusting number of batches to account for pattern overlap"<<endl;
  bool done(false);
  auto nbatch = max(minNBatches, nbatchEst);
  auto M(pat.size());

  //offset into text corresponding to start of each ring element
  vecInt offset(0); 
  size_t largestIx(0);
  while (!done){
    //try this combination of nbatch and M and adjust till it fits
    offset.clear();
    for(size_t bat = 0; bat < ringsize; bat++){
      offset.push_back(bat*(nbatch-M+1));
    }

    largestIx= offset[offset.size()-1]+(nbatch-1);

    if (largestIx >= txt.size()){
      done = true;
    } else {
      nbatch++;
      cout<<"increasing batch size to "<<nbatch<<endl;
    }
  }

  
  if (nbatch > maxNBatches) {
	cout<<"have to limit number of batches to "<<maxNBatches<<endl;
	nbatch = maxNBatches;
	if (largestIx+1 < txt.size()){
	  cout<<"have to limit text size to "<<largestIx+1<<endl;
	  txt.resize(largestIx+1);
	}
  }

  vecCT etxt(0);
  auto pt_len(0);  
  
  vecInt vin(0);
  for (size_t i = 0; i < nbatch; i++) {
    cout<<i<< '\r'<<flush;
    //build a vector out of the batches 
    for(size_t bat = 0; bat < ringsize; bat++){
      if (i+(unsigned)offset[bat] >= txt.size()) {
        vin.push_back('\0'); //null terminate
      } else {
        vin.push_back(txt[i+offset[bat]]);
      }
    }
    Plaintext pt= cc->MakePackedPlaintext(vin);
    pt_len = pt->GetLength();
    vin.clear();
    CT ct = cc->Encrypt(keyPair.publicKey, pt);
    etxt.push_back(ct);
  }
  cout<<"encrypted "<<etxt.size()<<" batches"<<endl;

  cout<<"Step 3.2 - Encrypt pattern"<<endl;  
  //encrypt the pattern
  // we copy the pattern over the entire ring
  vecCT epat(0);
  unsigned int j(0);
  for (auto ch: pat) {
	cout<<j<< '\r'<<flush;
	j++;
	for(size_t bat = 0; bat < ringsize; bat++){
	  vin.push_back(ch);
	}
	PT pt= cc->MakePackedPlaintext(vin);
	vin.clear();
	CT ct = cc->Encrypt(keyPair.publicKey, pt);
	epat.push_back(ct);
  }	

  cout<<"Step 4 - Encrypted string search"<<endl;  

  TIC(auto t2);

  vecCT eresult = encrypted_search(cc, keyPair.publicKey, epat, etxt, p);
  auto encrypted_time_ms = TOC_MS(t2);
  cout<< "Encrypted execution time "<<encrypted_time_ms<<" mSec."<<endl;  

  //decrypt the result and compute location of potential matches
  vecPT vecResult(0);
  for (auto e_itr:eresult){
	PT ptresult;  
	cc->Decrypt(keyPair.secretKey, e_itr, &ptresult);
	ptresult->SetLength(pt_len);
	vecResult.push_back(ptresult);
  }

  vecInt foundloc(0);
  for (size_t i = 0; i < vecResult.size(); i++) {
    auto unpackedVal = vecResult[i]->GetPackedValue();
    for (size_t jj = 0; jj< ringsize; jj++) {
      if (unpackedVal[jj] == 0) {
      auto loc = i + offset[jj];
      foundloc.push_back(loc);
      }
    }
  }

  sort(foundloc.begin(), foundloc.end());
  cout<<"total occurences "<<foundloc.size()<<endl;

  if (presult.size() != foundloc.size()) {
    cout<<"encrypted and plaintext results do not match"<<endl; 
    
    size_t smaller = min(presult.size(), foundloc.size());
    
    for (size_t i = 0; i < smaller; i++) {
      if (presult[i] != foundloc[i]) {
      cout <<"mismatch at location "<< i <<endl;
      }
    }      
    if (presult.size()<foundloc.size()) {
      for (size_t i = smaller; i< foundloc.size(); i++) {
      cout <<"encrypted extra finds "<<i<< foundloc[i]<<endl;
      }
    } else {
      for (size_t i = smaller; i< presult.size(); i++) {
      cout <<"plaintext extra finds "<<i<<" "<< presult[i]<<endl;
      }
    }
  } else {
	  cout<<"encrypted and plaintext results match"<<endl;
  }
  return 0;
}
